File: leetcode/509.fibonacci-number/cpp/fib.cpp
#include <array>

class Solution
{
public:
    int fib(int n)
    {
        if (n < 2)
            return n;

        for (int i = 2; i <= n; i++)
            dpFunc(i);

        return arr_[1];
    }

private:
    std::array<int, 2> arr_{0, 1};

    void dpFunc(int n)
    {
        auto v = arr_[0] + arr_[1];
        arr_[0] = arr_[1];
        arr_[1] = v;
    }
};

File: leetcode/509.fibonacci-number/ts/fib.ts
const fib = (n: number): number => {
    if (n < 2) return n

    const arr: number[] = Array.from({ length: 2 })
    
    const dpFunc = (n: number) => {
        const v = arr[0] + arr[1]
        arr[0] = arr[1]
        arr[1] = v
    }
    
    arr[0] = 0
    arr[1] = 1

    for (let i = 2; i <= n; i++) dpFunc(i)

    return arr[1]
}
